[prplMesh-dwpal] - Bring up DWPAL BWL mode

tomereli opened the pull request:
## Summary
Initial bring up for running prplMesh on Linux using DWPAL

Up till now, DUMMY mode which was the only supported mode so far on Linux. In Dummy mode, prplMesh agent is using stubbed BWL which doesn't interact with hostapd at all.
BWL DWPAL is the official mode supported currently only on Intel platforms - This PR provides initial support for running prplMesh agent over DWPAL, by adding example hostap.conf and an initial init script which insmods mac80211_hwsim.ko and starts hostapd. After this, the agent can start, attach to hostapd via DWPAL successfully, then fail on the first command it sends to hostapd - `GET_RADIO_INFO` - which is not supported by open source hostapd.

Next steps for prplMesh-dwpal full support:
- [ ] Build hostapd + Intel patches
- [ ] Update hostapd.conf based on Intel enhancements - specifically ACS_DONE mode
- [ ] Run prplMesh agent over dwpal, find the next gaps/issues and fix them. Eventuall with DWPAL and HWSIM, prplMesh should get to OPERATIONAL mode.
- [ ] Connect a client using wpa_cli and the other mac80211_hwsim radio, see it is shown in the connection map viw bml_dump_conn_map CLI command

## Build Instructions
### safeclib
Building safeclib is a prerequisite for building dwpal. This builds safelibc to the install folder to which we will install all the rest of the binaries/libraries.
```
cd <root>/3rdparty/safeclib
./build-tools/autogen.sh
./configure --prefix=$(realpath ../../)/build/install
make install
```
### dwpal
```
cd <root>/hostap/dwpal/build
cmake ../ -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=../../../build/install/ -G Ninja
ninja install
```
### prplMesh
```
cd <root>/prplMesh/tools
python3 ./maptools.py build all -f BWL_TYPE=DWPAL
```

## Testing
Test that the agent can connect to hostapd control socket via DWPAL and fail on GET_RADIO_INFO command.

* Initialize platform `sudo ./<root>/prplMesh/tools/scripts/prplMesh-platform-init.sh start`
* From a different shell. monitor the radio agent log `tail -F /tmp/$USER/beerocks/logs/beerocks_agent_wlan0.log`
* Start the agent `sudo ./<root>/build/install/bin/beerocks_agent`

Expected outcome seen in the log:
> NFO 13:59:32:359 <139901858465536> ap_manager_thread.cpp[347] --> waiting to attach to wlan0
INFO 13:59:32:759 <139901858465536> ap_manager_thread.cpp[347] --> waiting to attach to wlan0
INFO 13:59:33:160 <139901858465536> ap_manager_thread.cpp[347] --> waiting to attach to wlan0
DEBUG 13:59:33:560 <139901858465536> base_wlan_hal_dwpal.cpp[334] --> DWPAL FSM wlan0 Attach State: Init
INFO 13:59:33:560 <139901858465536> ap_manager_thread.cpp[347] --> waiting to attach to wlan0
DEBUG 13:59:33:961 <139901858465536> base_wlan_hal_dwpal.cpp[146] --> dwpal_hostap_interface_attach() success for: wlan0
INFO 13:59:33:961 <139901858465536> beerocks_state_machine.h[53] --> 'Init': Overriding destination for 'Attach' from 'Attach' to 'GetRadioInfo'
DEBUG 13:59:33:961 <139901858465536> base_wlan_hal_dwpal.cpp[334] --> DWPAL FSM wlan0 Attach State: GetRadioInfo
INFO 13:59:33:961 <139901858465536> ap_manager_thread.cpp[347] --> waiting to attach to wlan0
DEBUG 13:59:34:362 <139901858465536> base_wlan_hal_dwpal.cpp[420] --> 
cmd failed: GET_RADIO_INFO
DEBUG 13:59:34:363 <139901858465536> base_wlan_hal_dwpal.cpp[421] --> 
reply: UNKNOWN COMMAND

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tools: add linux platform init script

Initial platform init script for Linux which will
1. modprobe mac80211_hwsim
2. Start hostapd

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tools: add prplMesh.code-workspace

Add vscode workspace file for new directory structure:
prplMesh/
dwpal/
3rdparty/
	nng/
	safeclib/

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tools: move build folder outside of prplMesh repo

Before merging prplMesh repos into one repository, the build folder was
created in the top directory, where .repo folder exists.
This was done in order to prevent accidental removal when using git
clean -xdf, and to prevent having untracked folders.

Now that we have a single repository (prplMesh), the build folder is
created in prplMesh repo - so update build.py to create the folder in
the parent directory.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
bwl: add error print when dlopen fails

When building natively with BWL_TYPE=DWPAL, libdwpal might be installed
to a none system path, resulting with dlopen failure (setting
LD_LIBRARY_PATH resolves this).
Add an error when dlopen fails.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on tools/scripts/hostapd.conf:
```diff
@@ -0,0 +1,26 @@
+wpa=0
+bridge=br-lan
+ssid=dummy_ssid
+bssid=00:50:F1:20:D8:0E
```
It probably doesn't matter much, but wouldn't it be better to use a MAC address with the locally-administered bit set?

I typically use `22:01:02:03:04:05` and following.

arnout commented on tools/scripts/prplMesh-platform-init.sh:
```diff
@@ -0,0 +1,79 @@
+
+#!/bin/bash
```
If there is a line in front of it, the #! doesn't actually work.
It happens to "work" because when the kernel does not find a binfmt header (ELF header, a.out header, or #! header), it will fall back to `/bin/sh <file>` (legacy from ancient UNIX). So if /bin/sh happens to be bash in your case, it works.
Also, as it happens, there are no bashisms in the script, so /bin/sh works even if it is not bash. Well, there is one bashism: getopt - but the long options will just be ignored by non-bash shells I think.

arnout commented on tools/scripts/prplMesh-platform-init.sh:
```diff
@@ -0,0 +1,79 @@
+	echo -e '\033[1;31m'"$@"'\033[0m'
+}
+
+if [ `id -u` -ne 0 ]; then
```
Small nit: it would be nice if this would not be checked if you run it with -h.

tomereli replied:
done, also moved the check so that status command can run as none root

arnout commented on tools/scripts/prplMesh-platform-init.sh:
```diff
@@ -0,0 +1,79 @@
+
+VERBOSE=false
+HELP=false
+HOSTAPD=`which hostapd`
```
Redirect error to /dev/null. I don't have hostapd in my $PATH, so this prints an error even if I pass the -H option.
```suggestion
HOSTAPD=`which hostapd 2>/dev/null`
```

tomereli replied:
Done

arnout commented on agent/src/beerocks/slave/ap_wlan_hal_loader.cpp:
```diff
@@ -51,6 +51,7 @@ bwl::ap_wlan_hal *ap_wlan_hal_create(std::string iface_name, bool acs_enabled,
         });
 
         if (!s_sharedLib) {
+            LOG(ERROR) << "Failed loading " << BWL_MODULE_NAME << " shared library using dlopen";
```
Is there any reason why this is dlopen'ed instead of just linked?

tomereli replied:
Because the original idea was to make them as plugins. To allow you to swap the bwl for example without recompiling the agent..

arnout commented on tools/scripts/prplMesh-platform-init.sh:
```diff
@@ -0,0 +1,79 @@
+HELP=false
+HOSTAPD=`which hostapd`
+HOSTAPD_CONF=/etc/hostapd/hostapd.conf
+
```
```suggestion

run() {
    echo "$*"
    "$@" || exit $?
}

```

arnout commented on tools/scripts/prplMesh-platform-init.sh:
```diff
@@ -0,0 +1,79 @@
+
+start_function() {
+    echo "$0: start"
+    echo modprobe mac80211_hwsim
```
```suggestion
    run modprobe mac80211_hwsim
```

arnout commented on tools/scripts/hostapd.conf:
```diff
@@ -0,0 +1,26 @@
+channel=1
+
+interface=wlan0
+ctrl_interface=/var/run/hostapd
```
It would be nice it we could only use local directory, e.g. the build directory.

tomereli replied:
I tried not to add coupling of this script and hostap.conf - you can think of hostap.conf as an example for now. We can change it later on once we play a bit with hostapd and dwpal, so currently, I'll leave it as is.

arnout commented on tools/scripts/prplMesh-platform-init.sh:
```diff
@@ -0,0 +1,79 @@
+    echo "usage: $(basename $0) {start|stop|status} [-hv] [-H|--hostapd <hostapd>] [-C|--hostapd-conf <hostapd.conf>]"
+}
+
+# MAIN
```
I like to put the main in a function too, so that the only thing you have globally is function definitons and a single call to `main "$@"`. It's so much of a habit that I don't remember why :-)

tomereli replied:
Done

arnout commented on tools/scripts/prplMesh-platform-init.sh:
```diff
@@ -0,0 +1,79 @@
+eval set -- "$OPTS"
+
+while true; do
+  case "$1" in
```
Inconsistent indentation: the rest of the file uses 4 spaces.

tomereli replied:
Fixed

arnout commented on tools/scripts/prplMesh-platform-init.sh:
```diff
@@ -0,0 +1,79 @@
+stop_function() {
+    echo "$0: stop"
+    rmmod mac80211_hwsim
+    killall hostapd
```
First kill hostapd, then rmmod, no?

arnout commented on tools/scripts/prplMesh-platform-init.sh:
```diff
@@ -0,0 +1,79 @@
+VERBOSE=false
+HELP=false
+HOSTAPD=`which hostapd`
+HOSTAPD_CONF=/etc/hostapd/hostapd.conf
```
```suggestion
HOSTAPD_CONF=${0%/*}/hostapd.conf
```

arnout reviewed (approved):
Minor remarks only, nothing blocking.

Unfortunately, hostapd doesn't start correctly for me. I think I did something wrong while building it.

Commit:
build: add cmake RPATH settings for subtrees

Until https://github.com/prplfoundation/prplMesh/pull/1 is merged, add
cmake RPATH settings for all subtrees which don't have it.

For detailed description of this change - see commit 71ab4f2

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

tomereli commented:
@arnout I'm going to try something new, if we don't like it we can revert and merge it differently. Basically after merging this PR anyone should be able to continue the work of bringing up hostap + Intel patches and running prplMesh in DWPAL mode. So i'm going to merge this and keep the merge commit  which I assume will keep the PR description in our history.
